### PR TITLE
PB-5061: Kubevirt - Phase 2 Populate Resources for VM Backup in stork

### PR DESCRIPTION
**Pull request by vsundarraj-px**
**What type of PR is this?**
> Uncomment only one and also add the corresponding label in the PR:
>bug
feature
>improvement
>cleanup
>api-change
>design
>documentation
>failing-test
>unit-test
>integration-test

**What this PR does / why we need it**:
This PR implements PB-5061 kubevirt - phase 2 populate includeResource for VM backup. 

**Does this PR change a user-facing CRD or CLI?**:

yes
This introduces new filed backupObjectType: VirtualMachine 

Screenshot 2024-01-24 at 7 19 01 AM

**Is a release note needed?**:

```release-note
Issue:
User Impact:
Resolution

```
no
**Does this change need to be cherry-picked to a release branch?**:

no

The CR otuput which gets populated with IncludeResources with configMap and PVC
Screenshot 2024-01-24 at 7 19 40 AM

### CODE REVIEW

**Comment by vsundarraj-px** [comment]
Unit test case which create ruleCr and executing autoRules while creating VM backup.

File: pkg/applicationmanager/controllers/applicationbackup.go
Comment (siva-portworx) [inline pkg/applicationmanager/controllers/applicationbackup.go:268]: Better to add AlreadyExists error check and ignore the failure. Some time update might fail and we might re-enter again

Comment (vsundarraj-px) [reply pkg/applicationmanager/controllers/applicationbackup.go:268]: Added check for not found

File: pkg/applicationmanager/controllers/applicationbackup.go
Comment (vsundarraj-px) [reply pkg/applicationmanager/controllers/applicationbackup.go:2281]: fixed

File: pkg/applicationmanager/controllers/applicationbackup.go
Comment (vsundarraj-px) [reply pkg/applicationmanager/controllers/applicationbackup.go:2276]: Yes we have PB-5381

File: pkg/applicationmanager/controllers/applicationbackup.go
Comment (vsundarraj-px) [reply pkg/applicationmanager/controllers/applicationbackup.go:437]: Fixed

File: pkg/applicationmanager/controllers/applicationbackup.go
Comment (vsundarraj-px) [reply pkg/applicationmanager/controllers/applicationbackup.go:2300]: Since this is created even when applicationbackupCr is created outside of px-backup directly from stork. I kept it as stork.

**Comment by vsundarraj-px** [comment]
The PR is updated with changes that will address 
namespace and namespaceLabel VM specific backup request.  Output from backupCreate initiated from Automation. 
Screenshot 2024-01-30 at 5 43 53 PM (2)

**Comment by vsundarraj-px** [comment]
Screen shot of applicationbackupcr with namespace backup selecting VM resources in the namespace.

**Review by siva-portworx — APPROVED** [review]
LGTM

File: pkg/resourcecollector/virtualmachine.go
Comment (prashanthpx) [inline pkg/resourcecollector/virtualmachine.go:506]: From `GetVMIncludeListFromBackup()`  only those ns where VM is populated and reponse of vmlist if the final list with vm<-> ns mapping. under which condition would this if case be true?

File: pkg/applicationmanager/controllers/applicationbackup.go
Comment (vsundarraj-px) [reply pkg/applicationmanager/controllers/applicationbackup.go:97]: Updated.

File: pkg/applicationmanager/controllers/applicationbackup.go
Comment (vsundarraj-px) [reply pkg/applicationmanager/controllers/applicationbackup.go:450]: We will move to stageFinal if we do not update hence we need update when we fail as well.

File: pkg/applicationmanager/controllers/applicationbackup.go
Comment (vsundarraj-px) [reply pkg/applicationmanager/controllers/applicationbackup.go:2243]: This is logged in log.ApplicationBackupLog(backup) by the caller in hande().. Here we are just forming error string and returning..

File: pkg/resourcecollector/virtualmachine.go
Comment (vsundarraj-px) [reply pkg/resourcecollector/virtualmachine.go:391]: Thanks for the feedback.. old habbits, updated.

File: pkg/resourcecollector/virtualmachine.go
Comment (vsundarraj-px) [reply pkg/resourcecollector/virtualmachine.go:400]: This how this function is called.
getVmsFromNamespaceList(**backup.Spec.Namespaces**, objectMap)

In applicationbackup.go handle() function in initStage for labelselector we fetch all the namespace and update 
**backup.Spec.Namespaces** so our source of truth is **backup.Spec.Namespaces**
https://github.com/libopenstorage/stork/blob/a0782aa924765dea5746fde0b2d07239fdbb3aa0/pkg/applicationmanager/controllers/applicationbackup.go#L315

File: pkg/resourcecollector/virtualmachine.go
Comment (vsundarraj-px) [reply pkg/resourcecollector/virtualmachine.go:506]: We would have mapped VMs while processing backup.Spec.IncludeResources. But for VMs in namespaces will be differrred. So we map it here.
I have added comment for this now. Thanks.

**Review by prashanthpx — APPROVED** [review]

File: pkg/applicationmanager/controllers/applicationbackup.go
Comment (vsundarraj-px) [reply pkg/applicationmanager/controllers/applicationbackup.go:2242]: Yes for delete, there is PR https://github.com/libopenstorage/stork/pull/1616  raised which is blocked on this merge.

Comment (vsundarraj-px) [reply pkg/applicationmanager/controllers/applicationbackup.go:2242]: Noted on test case suggested.

File: pkg/applicationmanager/controllers/applicationbackup.go
Comment (vsundarraj-px) [reply pkg/applicationmanager/controllers/applicationbackup.go]: @prashanthpx I have updated to updateCr with right state. Please verify once before I can merge. Thanks again for identifying this.

File: pkg/applicationmanager/controllers/applicationbackup.go
Comment (dbinnal-px) [inline pkg/applicationmanager/controllers/applicationbackup.go:2286]: typo  'IsBackupObjectTypeVirtualMachine'

File: pkg/applicationmanager/controllers/applicationbackup.go
Comment (prashanthpx) [reply pkg/applicationmanager/controllers/applicationbackup.go]: Yes CR needs to be in FailedState.

File: pkg/applicationmanager/controllers/applicationbackup.go
Comment (vsundarraj-px) [reply pkg/applicationmanager/controllers/applicationbackup.go]: Fixed

**Pull request merged**
